csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -645,7 +645,86 @@ void IList.Remove(object value)
                 Remove(t);
             }
         }
-        #endregion        
+        #endregion
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public Enumerable AsStructEnumerable() => new Enumerable(this);
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly struct Enumerable : IEnumerable<T>
+        {
+            private readonly RepeatedField<T> field;
+
+            internal Enumerable(RepeatedField<T> field)
+            {
+                this.field = field;
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            /// <returns></returns>
+            public Enumerator GetEnumerator() => new Enumerator(field);
+
+            IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly RepeatedField<T> field;
+            private int index;
+
+            internal Enumerator(RepeatedField<T> field)
+            {
+                this.field = field;
+                this.index = -1;
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            public T Current => index == -1 || index >= field.Count
+                ? throw new InvalidOperationException()
+                : field[index];
+
+            object IEnumerator.Current => Current;
+
+            /// <summary>
+            /// 
+            /// </summary>
+            public void Dispose()
+            {
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            /// <returns></returns>
+            public bool MoveNext()
+            {
+                var count = field.Count;
+                index = Math.Min(index + 1, count);
```

Comment:
Why this extra check, instead of incrementing index always? Why do we care what is index value once it's equal to or greater than count? If the collection gets modified while iterating we don't need to guarantee any behaviour I guess.

Reply:
Overflow - calling MoveNext 2^32 times shouldn't make it valid again. Maybe that's slightly silly, but...

Reply:
I don't think the check is going to make much of a difference. If this turns out to be a performance problem, that would probably also go for the bounds checks on the array access. So it seems fine to keep it for now, for correctness sake.